### PR TITLE
Refine Jacoco plugin support for the configuration cache

### PR DESCRIPTION
**Pull request by eskatos**
Follow up to https://github.com/gradle/gradle/pull/14630

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java [hunk 1]
```diff
@@ -122,38 +121,55 @@ public void setReportsDir(File reportsDir) {
...
+            // Do not cache Test task if Jacoco doesn't produce its output as files
+            extension.isEnabled() && extension.getOutput() != JacocoTaskExtension.Output.FILE
+        );
+        task.getOutputs().doNotCacheIf(
+            "JaCoCo configured to not produce its output as a file",
+            element -> doNotCachePredicate.get()
+        );
+    }
+
+    private static class JacocoOutputCleanupTestTaskAction implements Action<Task> {
```
Comment (eskatos) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:149]: Using a serializable lambda here didn't work.

image

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java [hunk 2]
```diff
@@ -144,44 +145,17 @@ private void configureJacocoReportsDefaults(final JacocoPluginExtension extensio
...
-
-        private ReportOutputDirectoryAction(ProviderFactory providers, JacocoReport reportTask, Provider<File> reportsDir) {
-            this.providers = providers;
-            this.reportTask = reportTask;
-            this.reportsDir = reportsDir;
-        }
-
-        @Override
-        public void execute(ConfigurableReport report) {
+        reportTask.getReports().all((Action<ConfigurableReport> & Serializable) report ->
```
Comment (bamboo) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java:148]: 💄 `action(report -> ...)` where `action` is `SerializableLambdas.action`.

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java [hunk 3]
```diff
@@ -208,40 +182,21 @@ private void addDefaultReportTask(final JacocoPluginExtension extension, final T
                 reportTask.executionData(testTaskProvider.get());
                 reportTask.sourceSets(project.getExtensions().getByType(SourceSetContainer.class).getByName("main"));
                 // TODO: Change the default location for these reports to follow the convention defined in ReportOutputDirectoryAction
-                reportTask.getReports().all(new TestTaskReportOutputDirectoryAction(project.getProviders(), reportTask, project.provider(extension::getReportsDir), testTaskName));
+                Provider<File> reportsDir = project.provider(extension::getReportsDir);
+                reportTask.getReports().all((Action<ConfigurableReport> & Serializable) report -> {
```
Comment (bamboo) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java:186]: 💄 `action(report -> ...)`

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java [hunk 4]
```diff
@@ -122,38 +121,55 @@ public void setReportsDir(File reportsDir) {
         final String taskName = task.getName();
         LOGGER.debug("Applying Jacoco to " + taskName);
         final JacocoTaskExtension extension = task.getExtensions().create(TASK_EXTENSION_NAME, JacocoTaskExtension.class, objects, agent, task);
-        extension.setDestinationFile(providerFactory.provider(new Callable<File>() {
+        extension.setDestinationFile(providers.provider(new Callable<File>() {
```
Comment (bamboo) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:124]: 💄 `map` over `layout.getBuildDirectory`?

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java [hunk 5]
```diff
@@ -122,38 +121,55 @@ public void setReportsDir(File reportsDir) {
...
+            providers.provider(extension::getDestinationFile)
+        ));
 
         // Do not cache the task if we are not writing execution data to a file
-        task.getOutputs().doNotCacheIf("JaCoCo configured to not produce its output as a file", new Spec<Task>() {
-            @Override
-            public boolean isSatisfiedBy(Task element) {
-                // Do not cache Test task if Jacoco doesn't produce its output as files
-                return extension.isEnabled() && extension.getOutput() != JacocoTaskExtension.Output.FILE;
+        Provider<Boolean> doNotCachePredicate = providers.provider(() ->
```
Comment (bamboo) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:139]: 💄 same predicate is used above in `task.doFirst`, could reuse.

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java (hunk 1 above)
Comment (bamboo) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:149]: It seems it was capturing `this` (the extension) at that point. Maybe a variable declaration was missing?

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java (hunk 5 above)
Comment (eskatos) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:139]: It's not the same predicate

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java (hunk 2 above)
Comment (eskatos) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java:148]: Oh I missed that `action(..)` helper, nice! 👍

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java (hunk 1 above)
Comment (eskatos) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:144]: There was a reference to an outer field for `FileSystemOperations`. Once assigned to a local var the configuration cache is happy (and intellij says the var is redundant of course).

But then the up_to_dateness / caching behavior of the task is broken by the lambda. I'll leave it with the static class.

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java (hunk 5 above)
Comment (bamboo) [reply subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java:134]: Got it.

**Review by bamboo — APPROVED** [review]

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Pull request merged**